tests/apollo/test_skvbc_checkpoints.py

```diff
@@ -128,23 +127,69 @@ async def test_checkpoint_propagation_after_crashed_replicas_comeback(self, bft_
+        2) Crash f replicas, excluding the current primary
+        3) Send sufficient number of client requests to trigger checkpoint protocol
+        4) Make sure checkpoint is created
+        5) Bring the f replicas up
+        6) Repeat steps 2 - 5 nb_crashes number of times
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        n = bft_network.config.n
+        f = bft_network.config.f
+        c = bft_network.config.c
+
+        self.assertEqual(len(bft_network.procs), n,
+                         "Make sure all replicas are up initially.")
         current_primary = await bft_network.get_current_primary()
 
-        crashed_replicas = set()
+        await self._create_checkpoints_while_crashing(bft_network, skvbc, f, c,
+                                                      current_primary, nb_crashes=2)
+
+    async def _crash_replicas(
+            self, bft_network, nb_crashing, exclude_replicas=None):
+        crash_replicas = bft_network.random_set_of_replicas(nb_crashing, without=exclude_replicas)
+
+        bft_network.stop_replicas(crash_replicas)
+
+        return crash_replicas
+
+    async def _create_checkpoints_while_crashing(self,
+                                                 bft_network,
+                                                 skvbc, f, c,
+                                                 current_primary,
+                                                 nb_crashes=20):
+        for _ in range(nb_crashes):
+            crashed_replicas = await self._crash_replicas(
+                bft_network=bft_network,
+                nb_crashing=f,
+                exclude_replicas={current_primary}
+            )
+            self.assertFalse(current_primary in crashed_replicas)
+
+            self.assertGreaterEqual(len(bft_network.procs), 2 * f + c + 1,
+                                    "Make sure enough replicas are up to allow a successful checkpoint creation")
+
+            checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=current_primary)
+
+            await skvbc.fill_and_wait_for_checkpoint(
+                initial_nodes=bft_network.all_replicas(without=crashed_replicas),
+                checkpoint_num=1,
+                verify_checkpoint_persistency=False,
+                assert_state_transfer_not_started=False
+            )
+
+            await bft_network.wait_for_checkpoint(current_primary, expected_checkpoint_num=checkpoint_before + 1)
 
-        crash_candidates = bft_network.all_replicas(
-            without={current_primary})
-        random.shuffle(crash_candidates)
-        for i in range(nb_crashing):
-            bft_network.stop_replica(crash_candidates[i])
-            crashed_replicas.add(crash_candidates[i])
+            bft_network.start_replicas(crashed_replicas)
 
-        return crashed_replicas
+            await bft_network.wait_for_replicas_to_checkpoint(crashed_replicas, checkpoint_before + 1)
```

Comment:
Instead of doing this, I'd suggest the following:
```
await skvbc.fill_and_wait_for_checkpoint(
                initial_nodes=bft_network.all_replicas(),
                checkpoint_num=1,
                verify_checkpoint_persistency=False,
                assert_state_transfer_not_started=False
            )
```
This should force the stale replicas to rejoin the consensus and participate in the next checkpoint.

Reply:
I'd suggest to close this PR in favor of enhancing `test_checkpoint_propagation_after_crashed_replicas_comeback`, to verify the checkpoint propagates to all crashed replicas after they are brought up.

Next up, you could consider adding some network partitioning while creating checkpoints. See the existing examples in https://github.com/vmware/concord-bft/blob/master/tests/apollo/util/bft_network_partitioning.py